### PR TITLE
add checks on snow states after application of LDAS increments; StieglitzSnow cleanup

### PR DESCRIPTION
**Pull request by gmao-rreichle**
Summary of changes:

- Adds checks on snow states after application of LDAS increments from MODIS snow cover assimilation in coupled land-atm system (`catch_incr.F90`).

- Adds FICE[n] exports from Catchment (supersedes #814)

- 0-diff cleanup and documentation of `StieglitzSnow.F90`: 
  - adds missing "implicit none" to StieglitzSnow module
  - improves handling and naming of some public/private constants (CATCH_SNOW_*)
  - changed interface and functionality of StieglitzSnow_relayer() for use in LDAS
  - revises interface of snowrt() to group arguments with matching intent; add missing intent statement
  - fixes indentation (**best to hide white-space differences when viewing changes!**)
  - removes obsolete (commented-out) lines of code and constants
  - use StieglitzSnow_calc_asnow() instead of writing out equation each time
  - adds "module procedure" interface for StieglitzSnow_calc_asnow() to facilitate scalar and vector inputs and outputs
  - adds "module procedure" interface for StieglitzSnow_calc_tpsnow() for scalar/vector inputs [but maintains different if condition thresholds used in the original scalar (get_tf0d) and vector (get_tf_nd) subroutines for now
  - renames some public StieglitzSnow subroutines for clarity
  - cleanup of module use statements related to StieglitzSnow
  - adds documentation
  - removes commented-out and obsolete copy of StieglitzSnow subroutines from `GEOS_LandIceGridComp.F90` to avoid confusion
  - replaces hard-wired number of snow layers ("3") with parameter in `GEOS_Catch[*]GridComp.F90`

- Fixes file permissions of `SurfParams.F90` and `GEOS_SurfaceGridComp.rc`.

- Removes obsolete variables and use statements (catchment[CN].F90, GEOS_LandIceGridComp.F90, lsm_routines.F90)

- Removes suspicious catchcn_calc_*() subroutines in catchmentCN.F90

---

Testing to date:
- tested successfully for GEOSldas on Oct 6 (after https://github.com/GEOS-ESM/GEOSgcm_GridComp/pull/834/commits/667edae6b72e0f8b8f31b8f65aa7ade28bb8e1d3) by @gmao-rreichle (requires eponymous GEOSldas https://github.com/GEOS-ESM/GEOSldas/pull/676);  tests with aggressive compilation have round-off differences 
- retested successfully for GEOSldas **Intel-only** on Oct 15 (after https://github.com/GEOS-ESM/GEOSgcm_GridComp/pull/834/commits/48ff3cf3606ea63edb0dbbcb50162b9cd8f961a8) by @gmao-rreichle

- fully tested by @biljanaorescanin on 16/17 Oct 2023
  - All AGCM tests passed. (AMIP, REPLAY, No Inc Replay and Regression)
  - GEOSldas tests passed except for "aggressive" tests, which failed the comparison with roundoff differences, as expected.

---

@rdkoster @biljanaorescanin @weiyuan-jiang @lcandre2 @saraqzhang

### CODE REVIEW

**Comment by gmao-rreichle** [comment]
Converting back to draft so I can finish the non-essential cleanup, now that we're not furloughed.

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOScatch_GridComp/GEOS_CatchGridComp.F90 [hunk 1]
```diff
@@ -5392,6 +5396,12 @@ subroutine Driver ( RC )
                          CAPAC, CATDEF, RZEXC, SRFEXC,                                                   &
                          GHTCNT, WESNN, HTSNNN, SNDZN  )
                     
+                    ! @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
+                    !
+                    ! NEED TO RE-DIAGNOSE SNOW TEMP, ASNOW, etc HERE !?!?!?!?
```
Comment (lcandre2) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOScatch_GridComp/GEOS_CatchGridComp.F90:5401]: In GEOSldas, ASNOW is generally re-diagnosed after the increments are applied both in the snow da and others, but notably not TEMP, Albedo, etc. Perhaps, I am misinterpreting, but it seems logical to do all here instead?

Comment (lcandre2) [reply GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOScatch_GridComp/GEOS_CatchGridComp.F90:5401]: In clsm_ensupd_upd_routines.F90

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOScatch_GridComp/catch_incr.F90 [hunk 2]
```diff
@@ -141,84 +155,161 @@ subroutine check_catch_progn( NTILES,                                    &
...
+                sndzn( kk,ii) = max(sndzn( kk,ii), 0.)
+                
+                ! adjust snow depth to ensure  min <= density <= max
+                
+                snow_dens = (wesnn(kk,ii)/asnow_tmp)/sndzn(kk,ii)
+                
+                snow_dens = min( snow_dens, StieglitzSnow_RHOMA )
+                snow_dens = max( snow_dens, CATCH_SNWALB_RHOFS  )
+                
+                sndzn(kk,ii) = (wesnn(kk,ii)/asnow_tmp)/snow_dens
```
Comment (lcandre2) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOScatch_GridComp/catch_incr.F90:248]: If density is altered and used to recalculate depth, should heat content get the same treatment?

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSlandice_GridComp/GEOS_LandIceGridComp.F90 [hunk 3]
```diff
@@ -56,20 +62,20 @@ module GEOS_LandiceGridCompMod
   real,    parameter :: ALHM     = MAPL_ALHF   ! J/kg 
   real,    parameter :: TF       = MAPL_TICE   ! K
   real,    parameter :: RHOW     = MAPL_RHOWTR ! kg/m^3
+
   real,    parameter :: RHOFRESH = 300.        ! kg/m^3  density of fresh snow
```
Comment (lcandre2) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSlandice_GridComp/GEOS_LandIceGridComp.F90:66]: Updating these hard coded values is now on my list.

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/Shared/StieglitzSnow.F90 [hunk 4]
```diff
@@ -25,1357 +25,1439 @@ module StieglitzSnow
...
-       sndzprec = 0.
-       sndz1perc = 0.
-       wesnperc = 0.
-       wesndens = 0.
-       wesnrepar = 0.
-       wesnbot = 0.
-       !tksno   = condice
-       dtss    = 0. 
-       excswe  = 0.
+    !      integer, parameter :: nlay = 3             ! number of layers
```
Comment (lcandre2) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/Shared/StieglitzSnow.F90:286]: Is it worth removing 286-289?

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/Shared/StieglitzSnow.F90 [hunk 5]
```diff
@@ -25,1357 +25,1439 @@ module StieglitzSnow
...
+          densfac=amax1(0., amin1(1., icedens/rhofs))
+          term=densfac*snfr*(sndz(i)*rhow-wesn(i)*fices(i))
+          
+          if(pre > term) then
+             pre = min(pre - term, wesn(i))
              do k=1,N_constit
-               rconstit(1,k)=rconstit(1,k)+areasc0*totdepos(k)*dts
-               enddo
-
-             call relayer2(N_snow, N_constit, topthick, thickdist, &
```
Comment (lcandre2) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/Shared/StieglitzSnow.F90:441]: no more relayer2!!

**Review by lcandre2 — COMMENTED** [review]
Thank you @gmao-rreichle. I just had a few questions. I am a little perturbed by get_tf0d being slightly different than *_calc_tpsnow, but understand modifications would be non-0diff.

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOScatch_GridComp/GEOS_CatchGridComp.F90 (hunk 1 above)
Comment (gmao-rreichle) [reply GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOScatch_GridComp/GEOS_CatchGridComp.F90:5401]: Thanks, @lcandre2.  This block is specific to reading the increments from a file in the weakly-coupled land-atm DAS (LADAS).  So whatever we do here **only** applies to the LADAS.  (Not sure if you meant to imply that adding checks here would also take care of what may be missing in GEOSldas.)  In any case, this is clearly something that needs more thought.  
The PR adds what I hope to be reasonable checks on the Catchment **prognostics** in the LADAS.  Whether we need to update the **diagnostics** depends on where in the sequence of steps the increments are applied.  I don't have an answer to this yet.  This may require another PR later.  Since the present PR includes so much snow model clean-up that is needed for unrelated PRs, we can't wait with this PR until the "add checks" element of the present PR is perfect.

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOScatch_GridComp/catch_incr.F90 (hunk 2 above)
Comment (gmao-rreichle) [reply GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOScatch_GridComp/catch_incr.F90:248]: The idea here is to **not** change the mass, so the heat content should remain unchanged.

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/Shared/StieglitzSnow.F90 (hunk 5 above)
Comment (gmao-rreichle) [reply GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/Shared/StieglitzSnow.F90:441]: @lcandre2: I just renamed it for clarity.  I assume this comment doesn't ask for any action.  Let me know if otherwise.

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOScatchCN_GridComp/Shared/catchmentCN.F90 [hunk 6]
```diff
@@ -109,16 +105,21 @@ MODULE CATCHMENT_CN_MODEL
...
   
   private
   
   public :: catchcn
-  public :: catchcn_calc_tsurf
-  public :: catchcn_calc_tsurf_excl_snow
-  public :: catchcn_calc_etotl
+!!  public :: catchcn_calc_tsurf
+!!  public :: catchcn_calc_tsurf_excl_snow
+!!  public :: catchcn_calc_etotl
```
Comment (gmao-rreichle) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOScatchCN_GridComp/Shared/catchmentCN.F90:122]: @gmao-jkolassa, @rdkoster: These CatchCN-specific subroutines must have been copied from the corresponding Catch-specific subroutines at some point during CatchCN development.  The calculations seem wrong to me for CatchCN.  Fortunately, the subroutines do not seem to be used anywhere.  I commented them out for now.  If you can confirm that they are wrong, it's best to remove them from the code altogether.  Please let me know, thanks

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/Shared/StieglitzSnow.F90 [hunk 7]
```diff
@@ -25,1333 +25,1352 @@ module StieglitzSnow
...
+    
+    select case (tileType)
+    case (MAPL_LANDICE)
+       call FindTargetThickDist_Landice(N_snow, sndz, targetthick, topthick, thickdist)
+    case default
+       topthick  = targetthick(1)
+       thickdist = targetthick(2:N_snow)     
+    end select
+    
+    call StieglitzSnow_relayer( N_snow, N_constit, topthick, thickdist, &
```
Comment (gmao-rreichle) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/Shared/StieglitzSnow.F90:921]: @rdkoster, @lcandre2:  Note that the call to relayer here is followed by a call to calc_tpsnow and a subsequent check/adjustment of htsnn (see the next ~25 lines).  If relayer requires such a check/adjustment, would it not be better to include the check/adjustment within relayer?  
If that makes sense, then a simple call to relayer could perhaps be used to "fix" the snowpack after analysis increments have been applied.  
In a nutshell, I'm looking for a snow subroutine that is similar to what calc_soil_moist accomplishes for soil moisture (via calls to rzequil() and partition()), which guarantees that the soil moisture prognostics (catdef, rzexc, srfexc) after an analysis update are mutually consistent and won't crash the model at the next time step.

Comment (lcandre2) [reply GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/Shared/StieglitzSnow.F90:921]: @gmao-rreichle, I agree, it would be ideal if the result of calling relayer provided a physically consistent set of data with all structural conditions met. I am pretty strongly on the side of incorporating the 'conditions are conserved' caluclations into relayer. It seems currently, if relayer is called outside Stieglitz, then as it stands, the outputs may not be consistent, which is definitely non-ideal.

**Review by biljanaorescanin — APPROVED** [review]

**Comment by biljanaorescanin** [comment]
@sdrabenh this PR is ready for your review. 
I did standard testing on it (amip, replay and zero increment replay) and all tests passed.

**Review by lcandre2 — APPROVED** [review]

**Review by sdrabenh — APPROVED** [review]
